### PR TITLE
Remove `call_configure_sharded_model` lifecycle property

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Part of #8722

Changes:
- Remove the property guarding the hook: The Trainer will always call `configure_sharded_model` in each of `fit` `validate` `test` etc. This avoids subtle side-effects from prior runs and makes the call order consistent
- In turn, we strongly recommend users to implement `configure_sharded_model` idempotently. The update to `TestFSDPModel` demonstrates how one can check if the layers are already wrapped with FSDP and return early if so. This is the same spirit of https://github.com/PyTorchLightning/pytorch-lightning/issues/8593 but in user-land

@SeanNaren 

### Does your PR introduce any breaking changes? If yes, please list them.

- Calls `configure_sharded_model` unconditionally
- This removes the property on the accelerator/training type plugin
- Removes support for `call_configure_sharded_model_hook` on the LightningModule (which is not officially part of the LightningModule API) 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py [hunk 1]
```diff
@@ -49,27 +49,29 @@ def test_fsdp_with_sharded_amp(device_count_mock, mock_cuda_available, tmpdir):
...
-        # configure sharded model
-        self.call_configure_sharded_model_hook = False
-        # for loading full state dict, we first need to create a new unwrapped model
-        # to load state dict and then wrapping
-        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+        self._init_model()
 
     def configure_sharded_model(self) -> None:
+        # the model is already wrapped with FSDP: no need to wrap again!
+        if isinstance(self.layer, FullyShardedDataParallel):
```
Comment (four4fish) [inline tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py:65]: Will thi always return False? self.layer is a dict right?

Comment (ananthsub) [reply tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py:65]: no, it's an nn.Sequential module
but L70 `self.layer = wrap(self.layer)` will make this an instance of `FullyShardedDataParallel`

File: tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py [hunk 2]
```diff
@@ -49,27 +49,29 @@ def test_fsdp_with_sharded_amp(device_count_mock, mock_cuda_available, tmpdir):
...
-            return
-        # resetting call_configure_sharded_model_hook attribute so that we could call
-        # configure sharded model
-        self.call_configure_sharded_model_hook = False
-        # for loading full state dict, we first need to create a new unwrapped model
-        # to load state dict and then wrapping
-        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+        self._init_model()
 
     def configure_sharded_model(self) -> None:
```
Comment (carmocca) [inline tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py:63]: Can you update the docstring for this function to mention the idempotency requirement?

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/hooks.py#L295-L306

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/hooks.py [hunk 3]
```diff
@@ -297,12 +297,8 @@ def configure_sharded_model(self) -> None:
...
         memory and initialization time.
 
-        The accelerator manages whether to call this hook at every given stage.
-        For sharded plugins where model parallelism is required, the hook is usually on called once
-        to initialize the sharded parameters, and not called again in the same process.
-
-        By default for accelerators/plugins that do not use model sharding techniques,
-        this hook is called during each fit/val/test/predict stages.
+        This hook is called during each of fit/val/test/predict stages in the same process, so ensure that
+        implementation of this hook is idempotent.
```
Comment (ananthsub) [inline pytorch_lightning/core/hooks.py:301]: @carmocca do you think it'd be helpful to provide an example in case people are unfamiliar with this. such as, "For example, return early if your model is already sharded" ?

Comment (carmocca) [reply pytorch_lightning/core/hooks.py:301]: sure!

**Review by SeanNaren — APPROVED** [review]
Multiple calls to `configure_sharded_model` should be fine for DeepSpeed as well as there is a guard in place to not re-partition parameters that have already been partitioned: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/partition_parameters.py#L499

@ananthsub is it important to update the docs now? or should we wait till further changes are made to support a persistent model state across Trainer stages?

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by tchaton** [comment]
Hey @ananthsub,

Any progress on the failing Azure tests ?

Best,
T.C

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -401,20 +401,6 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
         """
         self.training_type_plugin.save_checkpoint(checkpoint, filepath)
 
-    @property
-    def call_configure_sharded_model_hook(self) -> bool:
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:405]: consider deprecation as it used to be public API
on the other hand, Accelerators are meant to be for internal use only...
cc: @awaelchli

**Comment by ananthsub** [comment]
> Hey @ananthsub,
> 
> Any progress on the failing Azure tests ?
> 
> Best,
> T.C

@tchaton I'm unable to reproduce the failures locally. is there any interleaving of tests that could cause this to fail in CI but not locally?

**Comment by awaelchli** [comment]
@ananthsub I can reproduce locally by running `py.test -v tests/plugins`. 

the first test that fails seems to be `test_deepspeed_skip_backward_raises`. 

When calling the tests individually, they pass. This is not unfamiliar as we had this a few times before. Since distributed logic is all globally shared in the e.g. in the torch package, things can leak from one test to the other (is my interpretation). 

```bash
tests/plugins/test_sharded_plugin.py::test_configure_ddp FAILED                                                                                                                                                                                                                                                                       [ 77%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded[DDPShardedPlugin] FAILED                                                                                                                                                                                                                                             [ 77%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded[DDPSpawnShardedPlugin] FAILED                                                                                                                                                                                                                                        [ 78%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded_reduce_buffer_size[1-params0-0] FAILED                                                                                                                                                                                                                               [ 78%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded_reduce_buffer_size[1-params1-128] FAILED                                                                                                                                                                                                                             [ 79%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded_reduce_buffer_size[2-params0-0] FAILED                                                                                                                                                                                                                               [ 80%]
tests/plugins/test_sharded_plugin.py::test_custom_kwargs_sharded_reduce_buffer_size[2-params1-128] FAILED                                                                                                                                                                                                                             [ 80%]
tests/plugins/test_sharded_plugin.py::test_block_backward_sync PASSED                                                                                                                                                                                                                                                                 [ 81%]
tests/plugins/test_single_device_plugin.py::test_single_cpu PASSED                                                                                                                                                                                                                                                                    [ 81%]
tests/plugins/test_single_device_plugin.py::test_single_gpu FAILED
```

Looking at the error messages, I see something disturbing:

```
tests/plugins/test_sharded_plugin.py:303:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pytorch_lightning/plugins/training_type/sharded.py:46: in configure_ddp
    LightningShardedDataParallel(self.model),
../../anaconda3/envs/pl/lib/python3.8/site-packages/deepspeed/runtime/zero/partition_parameters.py:240: in wrapper
    print_rank_0(f'Before initializing {module.__class__.__name__}',
```

The error message is originating from deepspeed 🤣 No clue how that can be. Need to investigate the code path.

**Comment by ananthsub** [comment]
> This is not unfamiliar as we had this a few times before. Since distributed logic is all globally shared in the e.g. in the torch package, things can leak from one test to the other (is my interpretation).

Do you remember how this was solved before?

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1276,18 +1276,9 @@ def _call_setup_hook(self) -> None:
...
-        # used on the model if the user re-create a trainer with resume_from_checkpoint
-        model = self.lightning_module
-        model_call_configure_sharded_model_hook = getattr(model, "call_configure_sharded_model_hook", False)
-        if self.accelerator.call_configure_sharded_model_hook and not model_call_configure_sharded_model_hook:
-            with self.accelerator.model_sharded_context():
-                self.call_hook("configure_sharded_model")
-                self.call_hook("on_configure_sharded_model")
-            model.call_configure_sharded_model_hook = True
-            self.accelerator.call_configure_sharded_model_hook = False
+        with self.accelerator.model_sharded_context():
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1279]: what I found so far is that this call to the model_sharded_context is causing the issue if it gets called for the deepspeed plugin. If I add 

```python
        if isinstance(self.training_type_plugin, DeepSpeedPlugin):
            return
```
before the call here, I don't see unrelated tests failing. 
So, investigating `DeepSpeedPlugin.model_sharded_context()`

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1279]: @ananthsub possible that I found the test that causes this. Can you try to got to the test
`test_deepspeed_multigpu_stage_3_warns_resume_training` in` tests/plugins/test_deepspeed_plugin`
and flip the flag `special=False` to `special=True`
?

**Comment by carmocca** [comment]
> Do you remember how this was solved before?

There is no solution without something like #8080. The workaround is to run the test per-process (what `special_tests.sh` does) and/or unrolling parametrizations

File: pytorch_lightning/accelerators/accelerator.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:405]: that's ok imo. the hook is not really valuable as a way to extend the plugins and it really only was a hack for the sharded integration :)
